x/icacallbacks/keeper/keeper.go

```diff
@@ -0,0 +1,125 @@
+		cdc          codec.BinaryCodec
+		storeKey     sdk.StoreKey
+		memKey       sdk.StoreKey
+		paramstore   paramtypes.Subspace
+		scopedKeeper capabilitykeeper.ScopedKeeper
+		icacallbacks map[string]types.ICACallbackHandler
+		IBCKeeper    ibckeeper.Keeper
+		ICAControllerKeeper   icacontrollerkeeper.Keeper
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey,
+	memKey sdk.StoreKey,
+	ps paramtypes.Subspace,
+	scopedKeeper capabilitykeeper.ScopedKeeper,
+	ibcKeeper ibckeeper.Keeper,
+	icacontrollerkeeper icacontrollerkeeper.Keeper,
+) *Keeper {
+	// set KeyTable if it has not already been set
+	if !ps.HasKeyTable() {
+		ps = ps.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		cdc:          cdc,
+		storeKey:     storeKey,
+		memKey:       memKey,
+		paramstore:   ps,
+		scopedKeeper: scopedKeeper,
+		icacallbacks: make(map[string]types.ICACallbackHandler),
+		IBCKeeper: ibcKeeper,
+		ICAControllerKeeper:   icacontrollerkeeper,
+	}
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// Should we add a `AddICACallback`
+func (k *Keeper) SetICACallbackHandler(module string, handler types.ICACallbackHandler) error {
+	_, found := k.icacallbacks[module]
+	if found {
+		return fmt.Errorf("callback handler already set for %s", module)
+	}
+	k.icacallbacks[module] = handler.RegisterICACallbacks()
+	return nil
+}
+
+func (k *Keeper) GetICACallbackHandler(module string) (types.ICACallbackHandler, error) {
+	callback, found := k.icacallbacks[module]
+	if !found {
+		return nil, fmt.Errorf("no callback handler found for %s", module)
+	}
+	return callback, nil
+}
+
+// ClaimCapability claims the channel capability passed via the OnOpenChanInit callback
+func (k *Keeper) ClaimCapability(ctx sdk.Context, cap *capabilitytypes.Capability, name string) error {
+	return k.scopedKeeper.ClaimCapability(ctx, cap, name)
+}
+
+func (k Keeper) CallRegisteredICACallback(ctx sdk.Context, modulePacket channeltypes.Packet, acknowledgement []byte) error {
+	// get the relevant module from the channel and port
+	portID := modulePacket.GetSourcePort()
+	channelID := modulePacket.GetSourceChannel()
+	module, _, err := k.IBCKeeper.ChannelKeeper.LookupModuleByChannel(ctx, portID, channelID)
+	if err != nil {
+		return err
+	}
+	// fetch the callback data
+	callbackDataKey := types.PacketID(portID, channelID, modulePacket.Sequence)
+	callbackData, found := k.GetCallbackData(ctx, callbackDataKey)
+	if !found {
+		errMsg := fmt.Sprintf("callback data not found for portID: %s, channelID: %s, sequence: %d", portID, channelID, modulePacket.Sequence)
+		k.Logger(ctx).Info(errMsg)
+		return nil
+	}
+	// fetch the callback function
+	callbackHandler, err := k.GetICACallbackHandler(module)
+	if err != nil {
+		k.Logger(ctx).Info("CallRegisteredICACallback", "err", err)
+		return err
+	}
+	// call the callback
+	if callbackHandler.HasICACallback(callbackData.CallbackId) {
```

Comment:
Great catch! I could see us either
- failing the call if a an ICA is executed with a callback that doesn't exist or
- removing the callback data here (even though no code is executed)

What do you think?

Reply:
Maybe move the call to `k.RemoveCallbackData()` inside this conditional so we remove iff one exists

Reply:
Thinking about this a bit more, I think the current behavior is correct (removing the callback data regardless of whether the callback is registered). `HasICACallback` checks if a callback function is registered, whereas `RemoveCallbackData` removes the bytes that are associated with a particular packet; if the callback caller tries calling a non-existent callback, we'd still want to remove that data to avoid state bloat